### ISSUE TITLE
Ractive.set calling observers as callbacks or deferred?

### ISSUE THREAD

**Issue by alexmakeev** [ractivejs/ractive#2016]
Dear Colleagues,

I honestly tried to dig the sources, but couldn't clarify the question.
Could any body guide me on how observers are being called by Ractive.set?
I need to catch a moment when all observers have already been processed (including semi-recursive one-after-another calls). I'm writing an adaptor with backbone-like data storage and want to be able to save data storage when form has been submitted (so, I can bind the Ractive.set(...).then() point, but it seems to be fully impossible to be sure that my observers, which I registered at the Adaptor wrapper, have been completely processed and the data at data storage has been synchronized with ractive form data.

For now the only alternative I see is to re-implement Ractive.set() method for all ractive instances to explicitly use my own implementation of DeferredObservers ( based on jquery.Deferred() ).

Thank you for any effort,
Alexander

**Comment by Skyrpex** [comment]
I suggest you to use $.Callbacks for that purpose. Promises will be called only once (like Ractive.observeOnce).

Also, Ractive.observe actually returns an object with a cancel method, so there's no room for a promise.

**Comment by fskreuz** [comment]
You might want to put up a simple demo, a jsfiddle perhaps.

Anyways, if you're curious as to how the internals work, you could cook up a simple script which breaks in various places. Something like the following. 

```js
data:{
  foo: '',
},
oninit: function(){
  this.observe('foo', function(){
    debugger; // If you arrive here, check the stack trace of operations.
  });
},
start: function(){
  debugger; // should start here
  this.set('foo','bar'); // use F11 to step into set. Use F8 to see which debugger it ends up
  debugger; // I believe you should hit the observer before arriving here
  this.set('foo','bar');
  debugger;
}
```

Open the dev tools and create a simple script that goes like this. Call the `start` method and you can step line-by-line in the debugger how the operations work.

**Comment by evs-chris** [comment]
0.8 had a good deal of cleanup on the runloop cycle that powers the promises returned by `set` and friends. There were some types of modification that were slipping out undetected. I'd say if `ractive.set(...).then(...)` isn't catching evereything, then it's definitely a bug that needs a new issue :smile:
